streams/src/test/java/org/apache/kafka/streams/kstream/RepartitionTopicNamingTest.java

```diff
@@ -0,0 +1,579 @@
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Properties;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+public class RepartitionTopicNamingTest {
+
+    private final KeyValueMapper<String, String, String> kvMapper = (k, v) -> k + v;
+    private static final String INPUT_TOPIC = "input";
+    private static final String COUNT_TOPIC = "outputTopic_0";
+    private static final String AGGREGATION_TOPIC = "outputTopic_1";
+    private static final String REDUCE_TOPIC = "outputTopic_2";
+    private static final String JOINED_TOPIC = "outputTopicForJoin";
+
+    private final String firstRepartitionTopicName = "count-stream";
+    private final String secondRepartitionTopicName = "aggregate-stream";
+    private final String thirdRepartitionTopicName = "reduced-stream";
+    private final String fourthRepartitionTopicName = "joined-stream";
+
+
+    @Test
+    public void shouldReuseFirstRepartitionTopicNameWhenOptimizing() {
+
+        final String optimizedTopology = buildTopology(StreamsConfig.OPTIMIZE).describe().toString();
+        final String unOptimizedTopology = buildTopology(StreamsConfig.NO_OPTIMIZATION).describe().toString();
+        final Pattern repartitionTopicPattern = Pattern.compile("Sink: .*-repartition");
+
+        assertThat(optimizedTopology, is(EXPECTED_OPTIMIZED_TOPOLOGY));
+        // only one repartition topic
+        assertThat(1, is(getCountOfRepartitionTopicsFound(optimizedTopology, repartitionTopicPattern)));
+        // the first named repartition topic
+        assertTrue(optimizedTopology.contains(firstRepartitionTopicName + "-repartition"));
+
+
+        assertThat(unOptimizedTopology, is(EXPECTED_UNOPTIMIZED_TOPOLOGY));
+        // now 4 repartition topic
+        assertThat(4, is(getCountOfRepartitionTopicsFound(unOptimizedTopology, repartitionTopicPattern)));
+        // all 4 named repartition topics present
+        assertTrue(unOptimizedTopology.contains(firstRepartitionTopicName + "-repartition"));
+        assertTrue(unOptimizedTopology.contains(secondRepartitionTopicName + "-repartition"));
+        assertTrue(unOptimizedTopology.contains(thirdRepartitionTopicName + "-repartition"));
+        assertTrue(unOptimizedTopology.contains(fourthRepartitionTopicName + "-left-repartition"));
+
+    }
+
+    // can't use same repartition topic name
+    @Test(expected = TopologyException.class)
+    public void shouldFailWithSameRepartitionTopicName() {
+        final StreamsBuilder builder = new StreamsBuilder();
+        builder.<String, String>stream("topic").selectKey((k, v) -> k).groupByKey(Grouped.named("grouping")).count().toStream();
+        builder.<String, String>stream("topicII").selectKey((k, v) -> k).groupByKey(Grouped.named("grouping")).count().toStream();
```

Comment:
Should this not fail here in `groupByKey()` already?

Maybe use `try-fail-catch` pattern here.

Reply:
I think the reason is that we do not check for unique names at Grouped / Joined, and hence only when later when the repartition topics are indeed going to be created the exception will be thrown. This looks fine to me.

Reply:
I personally highly prefer the try-fail-catch pattern, because it allows to narrow down which operation throws the exception. The current test would pass if the same exception is thrown one line above. IMHO, the `expected` annotation should only be used, if no other part in the test code could potentially throw the same exception (what is rarely the case).